src/EnergyPlus/PCMThermalStorage.cc

```diff
@@ -0,0 +1,629 @@
+                    approxTemp = Tmid;
+                    break;
+                }
+                if (hMid > targetEnthalpy) {
+                    Thigh = Tmid;
+                } else {
+                    Tlow = Tmid;
+                }
+            }
+
+            this->PCM_TankTemp = approxTemp;
+        }
+
+        // Determine operation mode.  The sign of the heat transfer on each side dictates whether the unit should
+        // charge or discharge.  In the original implementation a second condition on the tank temperature
+        // prevented charging unless the PCM temperature was below the melting point, which caused the plant
+        // side flow to be ignored even when the PCM should have been storing energy.  Here we rely solely
+        // on the direction of heat transfer to set the mode.
+        // SOC as fraction 0–1
+        Real64 soc = EnergyStored / (TankCapacity * LatentHeat);
+        soc = std::clamp(soc, 0.0, 1.0);
+
+        // Hysteresis thresholds (tune as needed)
+        constexpr Real64 cutInSOC = 0.40;  // start charging when SOC <= 40%
+        constexpr Real64 cutOutSOC = 0.95; // stop charging when SOC >= 95%
+
+        // Persistent mode
+        static bool chargingMode = false;
+        if (soc <= cutInSOC) {
+            chargingMode = true;
+        }
+        if (soc >= cutOutSOC) {
+            chargingMode = false;
+        }
+
+        // Decide flows: charge => plant flows only; discharge => use flows only
+        Real64 mUseReq = 0.0;
+        Real64 mPlantReq = 0.0;
+
+        if (chargingMode) {
+            mPlantReq = this->PlantSideMassFlowRate;
+        } else {
+            mUseReq = this->UseSideMassFlowRate;
+        }
+
+        // Issue flow requests (mass flow, kg/s)
+        PlantUtilities::SetComponentFlowRate(state, mUseReq, this->UseSideInletNode, this->UseSideOutletNode, this->usePlantLoc);
+        PlantUtilities::SetComponentFlowRate(state, mPlantReq, this->PlantSideInletNode, this->PlantSideOutletNode, this->sourcePlantLoc);
+
+        // Recompute heat-transfer rates using the requested flows (W)
+        // (deltaTUse/deltaTPlant were computed above; keep your existing outlet temp calcs)
+        Real64 useheatTransfer_req = mUseReq * CpWater * (useInlet.Temp - useOutletTemp);
+        Real64 plantheatTransfer_req = mPlantReq * CpWater * (plantInlet.Temp - plantOutletTemp);
+
+        // Only one side should contribute per timestep by construction; but compute net formally:
+        Real64 netPowerW = plantheatTransfer_req + useheatTransfer_req - HeatLossRate_W;
+
+        // Update stored energy (J)
+        EnergyStored += netPowerW * dt_seconds;
+
+        // Update node temps for the side that actually flowed
+        if (mUseReq > 0.0) {
+            useOutlet.Temp = useOutletTemp;
+        } else if (mPlantReq > 0.0) {
+            plantOutlet.Temp = std::min(plantOutletTemp, state.dataPlnt->PlantLoop[this->sourcePlantLoc.loopNum].MaxTemp);
```

Comment:
If the plant outlet temp gets clipped here then the calculation for plantheatTransfer_req, netPowerW and EnergyStored above are inaccurate.